jaxrs/src/main/java/org/killbill/billing/jaxrs/resources/AdminResource.java

```diff
@@ -104,4 +123,106 @@ public boolean apply(final PaymentTransaction input) {
+
+        // clear account-bcd cache by accountId
+        final Ehcache accountBcdCache = cacheManager.getEhcache(CacheType.ACCOUNT_BCD.getCacheName());
+        accountBcdCache.remove(UUID.fromString(accountId));
+
+        return Response.status(Status.OK).build();
+    }
+
+    @DELETE
+    @Path("/" + CACHE + "/" + TENANTS)
+    @Produces(APPLICATION_JSON)
+    @ApiOperation(value = "Invalidates Caches per tenant level")
+    @ApiResponses(value = {})
+    public Response invalidatesCacheByTenant(@QueryParam("tenantApiKey") final String tenantApiKey,
+                                              @javax.ws.rs.core.Context final HttpServletRequest request) throws TenantApiException {
+
+        // creating Tenant Context from Request
+        TenantContext tenantContext = context.createContext(request);
+
+        Tenant currentTenant = tenantApi.getTenantById(tenantContext.getTenantId());
+
+        // getting Tenant Record Id
+        Long tenantRecordId = recordIdApi.getRecordId(tenantContext.getTenantId(), ObjectType.TENANT, tenantContext);
+
+        // clear tenant-record-id cache by tenantId
+        final Ehcache tenantRecordIdCache = cacheManager.getEhcache(CacheType.TENANT_RECORD_ID.getCacheName());
+        tenantRecordIdCache.remove(currentTenant.getId().toString());
+
+        // clear tenant-payment-state-machine-config cache by tenantRecordId
+        final Ehcache tenantPaymentStateMachineConfigCache = cacheManager.getEhcache(CacheType.TENANT_PAYMENT_STATE_MACHINE_CONFIG.getCacheName());
+        removeCacheByKey(tenantPaymentStateMachineConfigCache, tenantRecordId.toString());
+
+        // clear tenant cache by tenantApiKey
+        final Ehcache tenantCache = cacheManager.getEhcache(CacheType.TENANT.getCacheName());
+        tenantCache.remove(currentTenant.getApiKey());
+
+        // clear tenant-kv cache by tenantRecordId
+        final Ehcache tenantKvCache = cacheManager.getEhcache(CacheType.TENANT_KV.getCacheName());
+        removeCacheByKey(tenantKvCache, tenantRecordId.toString());
+
+        // clear tenant-config cache by tenantRecordId
+        final Ehcache tenantConfigCache = cacheManager.getEhcache(CacheType.TENANT_CONFIG.getCacheName());
+        tenantConfigCache.remove(tenantRecordId);
+
+        // clear tenant-overdue-config cache by tenantRecordId
+        final Ehcache tenantOverdueConfigCache = cacheManager.getEhcache(CacheType.TENANT_OVERDUE_CONFIG.getCacheName());
+        tenantOverdueConfigCache.remove(tenantRecordId);
+
+        // clear tenant-catalog cache by tenantRecordId
+        final Ehcache tenantCatalogCache = cacheManager.getEhcache(CacheType.TENANT_CATALOG.getCacheName());
+        tenantCatalogCache.remove(tenantRecordId);
+
+        return Response.status(Status.OK).build();
+    }
+
+    private void removeCacheByKey(final Ehcache tenantCache, final String tenantRecordId) {
+        for (String key : (List<String>) tenantCache.getKeys()) {
```

Comment:
Nit: such casts don't always behave well at runtime and can hide some cryptic bugs. Although it's probably not an issue here, I would suggest rewriting it:

```
for (final Object key : tenantCache.getKeys()) {
  if (key != null && key.toString().endsWith("::" + tenantRecordId)) {
    ...
```